MAINT: port minpack2.dcsrch from Fortran to Python, remove Fortran code

andyfaff opened the pull request:
Currently `scipy.optimize.minpack2` is used in linesearch operations in `scipy.optimize._linesearch.line_search_wolfe1`.  minpack2 is a Fortran module that exports two functions, `dcsrch` and `dcstep`.

I have ported the contents of those functions to Python. I propose that we switch to using the Python port, as this will be easier to maintain than the Fortran version. The Python version is a straightforward translation of the Fortran.

I have examined the relative timings of the Fortran module with the Python version using the example at https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.line_search.html, and there is no difference. I do recognise this is a simple example. However, it's probable that most of the call time in a linesearch will be to the original objective function. In fact using a simple example like this should highlight any inefficiencies of the Python port itself.

```
import numpy as np
from scipy.optimize import line_search

def obj_func(x):
    return (x[0])**2+(x[1])**2
def obj_grad(x):
    return [2*x[0], 2*x[1]]

start_point = np.array([1.8, 1.7])
search_gradient = np.array([-1.0, -1.0])
```
```
# ported code
%timeit line_search(obj_func, obj_grad, start_point, search_gradient)
5.97 µs ± 31.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
# Fortran code
%timeit line_search(obj_func, obj_grad, start_point, search_gradient)
5.94 µs ± 19.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

The `dcsrch` code could be wrapped up even further into a standalone function, rather than create an object and call it. However, it's an internal function so it probably doesn't matter so much.

At the moment a `minpack2` module can be imported from `scipy.optimize`.  Its namespace is deprecated for removal in scipy 2. If I [search for minpack2 on github](https://github.com/search?q=minpack2.dcsrch&type=code) it seems reasonably common, with 131 hits.
In one of the commits I have removed minpack2 entirely (although there is a vendored section of the same code in lbfgsb). I have not compensated for its removal by pointing to the Python version. If we want to keep minpack2 around for longer I can back out one of the commits.

With regards to testing, it passes the test suite, and I have checked that both the Fortran and Python versions give the same output for the example given above. I have not attempted to write further tests.

ilayn commented:
Other than the intrinsic weirdness of the code coming from the Fortran side (for a lack of a better tool other than bytestrings), I'd say this looks good to me. I'd leave the deprecation strategy to @j-bowhay and @h-vetinari but `minpack2` was/should be internal anyways. 

Much to my surprise, the code is in fact simple and we can refactor it afterwards.

andyfaff commented:
I had a look at lbfgsb - that's truly weird. Labels and jumps everywhere.

ilayn commented:
Yes typical F77 code from '90s. I guess we'll get there when we get there with LBFGSB.

j-bowhay commented on scipy/optimize/_dcsrch.py:
```diff
@@ -0,0 +1,728 @@
+            task = b"FG"
+            return stp, f, g, task
+
+        else:
+            pass
```
Slightly confusing having this no op but guess it corresponds to the restore local variables section? Could we either remove or comment

andyfaff replied:
Yes, it corresponds to the restore local variables section. I thought for a first pass it would be a good idea to keep it like the original Fortran, to aid problem solving. I'll remove it, but add a comment.

j-bowhay commented on scipy/optimize/_dcsrch.py:
```diff
@@ -0,0 +1,728 @@
+    sgn_dp = np.sign(dp)
+    sgn_dx = np.sign(dx)
+
+    # sgnd = dp * (dx / abs(dx))
+    sgnd = sgn_dp * sgn_dx
```
Just to double-check my own reasoning, these are equivalent in sign but not value however this doesn't matter because of how they are used?

andyfaff replied:
In the Fortran code `sgnd` is used as:
```
else if (sgnd .lt. zero) then
```
and
```
         if (sgnd .lt. zero) then
```

so the logic expressions are asking is *`sgnd` positive or negative`*, it doesn't want the magnitude, just the sign.
From my reading of the code it wants to know if `dp` and `dx` have opposing signs, which is why it works out the product. 
The way to do that in Fortran must be to do the calculation that's commented out. However, if you calculate `dx/abs(dx)` in Python you get overflow errors if `dx` is infinity - from running the test suite this is a common occurrence. The easy way to fix that is just obtain the sign of both numbers.

j-bowhay reviewed (commented):
Thanks @andyfaff. I actually printed off the source code for this with the intention of having a go and since have not found the time! However, I did read through pretty much line by line and am happy the implementations match.

With regard to the deprecation, there are few active projects using `minpack2.dcsrch` eg. https://github.com/idaholab/raven. Annoyingly there isn't currently a deprecation warning for this:
```
In [2]: from scipy.optimize import minpack2
```
only:
```
In [1]: from scipy.optimize.minpack2 import dcsrch
<ipython-input-1-e15cc5999327>:1: DeprecationWarning: `scipy.optimize.minpack2.dcsrch` is deprecated along with the `scipy.optimize.minpack2` namespace. `scipy.optimize.minpack2.dcsrch` will be removed in SciPy 1.13.0, and the `scipy.optimize.minpack2` namespace will be removed in SciPy 2.0.0.
```
My preference would be to keep minpack2 around until there is at least a proper deprecation warning.

andyfaff commented:
I don't think we raise a warning on the modules themselves, just on the functions. I would point out that if you do:
```
from scipy.optimize import minpack2
minpack2.dcsrch
```
you don't get a DeprecationWarning from the module import itself, but you do get a DeprecationWarning when you go to use a function from it. So people should be experiencing warnings. Perhaps we should modify the DeprecationWarning code to bring forward the removal date for minpack2?

j-bowhay commented:
> I don't think we raise a warning on the modules themselves, just on the functions. I would point out that if you do:
> 
> ```
> from scipy.optimize import minpack2
> minpack2.dcsrch
> ```
> 
> you don't get a DeprecationWarning from the module import itself, but you do get a DeprecationWarning when you go to use a function from it. So people should be experiencing warnings. Perhaps we should modify the DeprecationWarning code to bring forward the removal date for minpack2?

Ah yes, you're right about the warnings I don't have an issue there then. The `__all__` for `minpack2` contains 
```
__all__ = [  # noqa: F822
    'dcsrch',
    'dcstep',
]
```
and both of these have warnings stating removal for 1.13. After this the module would be empty anyway so I think we could probably safely remove it at the same time

andyfaff commented:
<details><summary>The patch for the removal </summary>
<p>

```
diff --git a/scipy/_lib/tests/test_public_api.py b/scipy/_lib/tests/test_public_api.py
index 4446a50af..f760705cb 100644
--- a/scipy/_lib/tests/test_public_api.py
+++ b/scipy/_lib/tests/test_public_api.py
@@ -134,7 +134,6 @@ PRIVATE_BUT_PRESENT_MODULES = [
     'scipy.optimize.lbfgsb',
     'scipy.optimize.linesearch',
     'scipy.optimize.minpack',
-    'scipy.optimize.minpack2',
     'scipy.optimize.moduleTNC',
     'scipy.optimize.nonlin',
     'scipy.optimize.optimize',
@@ -332,7 +331,6 @@ def test_api_importable():
                           ('scipy.optimize.lbfgsb', None),
                           ('scipy.optimize.linesearch', None),
                           ('scipy.optimize.minpack', None),
-                          ('scipy.optimize.minpack2', None),
                           ('scipy.optimize.moduleTNC', None),
                           ('scipy.optimize.nonlin', None),
                           ('scipy.optimize.optimize', None),
diff --git a/scipy/optimize/__init__.py b/scipy/optimize/__init__.py
index 588a65b26..209328c96 100644
--- a/scipy/optimize/__init__.py
+++ b/scipy/optimize/__init__.py
@@ -435,7 +435,7 @@ from ._milp import milp
 
 # Deprecated namespaces, to be removed in v2.0.0
 from . import (
-    cobyla, lbfgsb, linesearch, minpack, minpack2, moduleTNC, nonlin, optimize,
+    cobyla, lbfgsb, linesearch, minpack, moduleTNC, nonlin, optimize,
     slsqp, tnc, zeros
 )
 
diff --git a/scipy/optimize/linesearch.py b/scipy/optimize/linesearch.py
index 0d1a04d83..7a3b39103 100644
--- a/scipy/optimize/linesearch.py
+++ b/scipy/optimize/linesearch.py
@@ -12,7 +12,6 @@ __all__ = [  # noqa: F822
     'line_search_armijo',
     'line_search_wolfe1',
     'line_search_wolfe2',
-    'minpack2',
     'scalar_search_armijo',
     'scalar_search_wolfe1',
     'scalar_search_wolfe2',
diff --git a/scipy/optimize/meson.build b/scipy/optimize/meson.build
index 001335c03..ff404d393 100644
--- a/scipy/optimize/meson.build
+++ b/scipy/optimize/meson.build
@@ -144,23 +144,6 @@ _cobyla = py3.extension_module('_cobyla',
   subdir: 'scipy/optimize'
 )
 
-minpack2_module = custom_target('minpack2_module',
-  output: ['_minpack2module.c'],
-  input: 'minpack2/minpack2.pyf',
-  command: [py3, generate_f2pymod, '@INPUT@', '-o', '@OUTDIR@']
-)
-
-minpack2 = py3.extension_module('_minpack2',
-  [minpack2_module, 'minpack2/dcsrch.f', 'minpack2/dcstep.f'],
-  c_args: numpy_nodepr_api,
-  fortran_args: fortran_ignore_warnings,
-  link_args: version_link_args,
-  dependencies: [fortranobject_dep],
-  install: true,
-  link_language: 'fortran',
-  subdir: 'scipy/optimize'
-)
-
 slsqp_module = custom_target('slsqp_module',
   output: ['_slsqpmodule.c'],
   input: 'slsqp/slsqp.pyf',
@@ -289,7 +272,6 @@ py3.install_sources([
   'lbfgsb.py',
   'linesearch.py',
   'minpack.py',
-  'minpack2.py',
   'moduleTNC.py',
   'nonlin.py',
   'optimize.py',
diff --git a/scipy/optimize/minpack2.py b/scipy/optimize/minpack2.py
deleted file mode 100644
index 6e961f424..000000000
--- a/scipy/optimize/minpack2.py
+++ /dev/null
@@ -1,21 +0,0 @@
-# This file is not meant for public use and will be removed in SciPy v2.0.0.
-# Use the `scipy.optimize` namespace for importing the functions
-# included below.
-
-from scipy._lib.deprecation import _sub_module_deprecation
-
-
-__all__ = [  # noqa: F822
-    'dcsrch',
-    'dcstep',
-]
-
-
-def __dir__():
-    return __all__
-
-
-def __getattr__(name):
-    return _sub_module_deprecation(sub_package="optimize", module="minpack2",
-                                   private_modules=["_minpack2"], all=__all__,
-                                   attribute=name)
diff --git a/scipy/optimize/minpack2/dcsrch.f b/scipy/optimize/minpack2/dcsrch.f
deleted file mode 100644
index fd4a154d6..000000000
--- a/scipy/optimize/minpack2/dcsrch.f
+++ /dev/null
@@ -1,349 +0,0 @@
-      subroutine dcsrch(stp,f,g,ftol,gtol,xtol,task,stpmin,stpmax,
-     +                  isave,dsave)
-      character*(*) task
-      integer isave(2)
-      double precision f, g, stp, ftol, gtol, xtol, stpmin, stpmax
-      double precision dsave(13)
-c     **********
-c
-c     Subroutine dcsrch
-c
-c     This subroutine finds a step that satisfies a sufficient
-c     decrease condition and a curvature condition.
-c
-c     Each call of the subroutine updates an interval with
-c     endpoints stx and sty. The interval is initially chosen
-c     so that it contains a minimizer of the modified function
-c
-c           psi(stp) = f(stp) - f(0) - ftol*stp*f'(0).
-c
-c     If psi(stp) <= 0 and f'(stp) >= 0 for some step, then the
-c     interval is chosen so that it contains a minimizer of f.
-c
-c     The algorithm is designed to find a step that satisfies
-c     the sufficient decrease condition
-c
-c           f(stp) <= f(0) + ftol*stp*f'(0),
-c
-c     and the curvature condition
-c
-c           abs(f'(stp)) <= gtol*abs(f'(0)).
-c
-c     If ftol is less than gtol and if, for example, the function
-c     is bounded below, then there is always a step which satisfies
-c     both conditions.
-c
-c     If no step can be found that satisfies both conditions, then
-c     the algorithm stops with a warning. In this case stp only
-c     satisfies the sufficient decrease condition.
-c
-c     A typical invocation of dcsrch has the following outline:
-c
-c     Evaluate the function at stp = 0.0d0; store in f.
-c     Evaluate the gradient at stp = 0.0d0; store in g.
-c     Choose a starting step stp.
-c
-c     task = 'START'
-c  10 continue
-c        call dcsrch(stp,f,g,ftol,gtol,xtol,task,stpmin,stpmax,
-c    +               isave,dsave)
-c        if (task .eq. 'FG') then
-c           Evaluate the function and the gradient at stp
-c           go to 10
-c           end if
-c
-c     NOTE: The user must not alter work arrays between calls.
-c
-c     The subroutine statement is
-c
-c       subroutine dcsrch(f,g,stp,ftol,gtol,xtol,stpmin,stpmax,
-c                         task,isave,dsave)
-c     where
-c
-c       stp is a double precision variable.
-c         On entry stp is the current estimate of a satisfactory
-c            step. On initial entry, a positive initial estimate
-c            must be provided.
-c         On exit stp is the current estimate of a satisfactory step
-c            if task = 'FG'. If task = 'CONV' then stp satisfies
-c            the sufficient decrease and curvature condition.
-c
-c       f is a double precision variable.
-c         On initial entry f is the value of the function at 0.
-c            On subsequent entries f is the value of the
-c            function at stp.
-c         On exit f is the value of the function at stp.
-c
-c       g is a double precision variable.
-c         On initial entry g is the derivative of the function at 0.
-c            On subsequent entries g is the derivative of the
-c            function at stp.
-c         On exit g is the derivative of the function at stp.
-c
-c       ftol is a double precision variable.
-c         On entry ftol specifies a nonnegative tolerance for the
-c            sufficient decrease condition.
-c         On exit ftol is unchanged.
-c
-c       gtol is a double precision variable.
-c         On entry gtol specifies a nonnegative tolerance for the
-c            curvature condition.
-c         On exit gtol is unchanged.
-c
-c       xtol is a double precision variable.
-c         On entry xtol specifies a nonnegative relative tolerance
-c            for an acceptable step. The subroutine exits with a
-c            warning if the relative difference between sty and stx
-c            is less than xtol.
-c         On exit xtol is unchanged.
-c
-c       task is a character variable of length at least 60.
-c         On initial entry task must be set to 'START'.
-c         On exit task indicates the required action:
-c
-c            If task(1:2) = 'FG' then evaluate the function and
-c            derivative at stp and call dcsrch again.
-c
-c            If task(1:4) = 'CONV' then the search is successful.
-c
-c            If task(1:4) = 'WARN' then the subroutine is not able
-c            to satisfy the convergence conditions. The exit value of
-c            stp contains the best point found during the search.
-c
-c            If task(1:5) = 'ERROR' then there is an error in the
-c            input arguments.
-c
-c         On exit with convergence, a warning or an error, the
-c            variable task contains additional information.
-c
-c       stpmin is a double precision variable.
-c         On entry stpmin is a nonnegative lower bound for the step.
-c         On exit stpmin is unchanged.
-c
-c       stpmax is a double precision variable.
-c         On entry stpmax is a nonnegative upper bound for the step.
-c         On exit stpmax is unchanged.
-c
-c       isave is an integer work array of dimension 2.
-c
-c       dsave is a double precision work array of dimension 13.
-c
-c     Subprograms called
-c
-c       MINPACK-2 ... dcstep
-c
-c     MINPACK-1 Project. June 1983.
-c     Argonne National Laboratory.
-c     Jorge J. More' and David J. Thuente.
-c
-c     MINPACK-2 Project. November 1993.
-c     Argonne National Laboratory and University of Minnesota.
-c     Brett M. Averick, Richard G. Carter, and Jorge J. More'.
-c
-c     **********
-      double precision zero, p5, p66
-      parameter (zero=0.0d0,p5=0.5d0,p66=0.66d0)
-      double precision xtrapl, xtrapu
-      parameter (xtrapl=1.1d0,xtrapu=4.0d0)
-
-      logical brackt
-      integer stage
-      double precision finit, ftest, fm, fx, fxm, fy, fym, ginit, gtest,
-     +                 gm, gx, gxm, gy, gym, stx, sty, stmin, stmax,
-     +                 width, width1
-
-      external dcstep
-
-c     Initialization block.
-
-      if (task(1:5) .eq. 'START') then
-
-c        Check the input arguments for errors.
-         
-         if (stp .lt. stpmin) task = 'ERROR: STP .LT. STPMIN'
-         if (stp .gt. stpmax) task = 'ERROR: STP .GT. STPMAX'
-         if (g .ge. zero) task = 'ERROR: INITIAL G .GE. ZERO'
-         if (ftol .lt. zero) task = 'ERROR: FTOL .LT. ZERO'
-         if (gtol .lt. zero) task = 'ERROR: GTOL .LT. ZERO'
-         if (xtol .lt. zero) task = 'ERROR: XTOL .LT. ZERO'
-         if (stpmin .lt. zero) task = 'ERROR: STPMIN .LT. ZERO'
-         if (stpmax .lt. stpmin) task = 'ERROR: STPMAX .LT. STPMIN'
-
-c        Exit if there are errors on input.
-
-         if (task(1:5) .eq. 'ERROR') return
-
-c        Initialize local variables.
-
-         brackt = .false.
-         stage = 1
-         finit = f
-         ginit = g
-         gtest = ftol*ginit
-         width = stpmax - stpmin
-         width1 = width/p5
-
-c        The variables stx, fx, gx contain the values of the step,
-c        function, and derivative at the best step.
-c        The variables sty, fy, gy contain the value of the step,
-c        function, and derivative at sty.
-c        The variables stp, f, g contain the values of the step,
-c        function, and derivative at stp.
-
-         stx = zero
-         fx = finit
-         gx = ginit
-         sty = zero
-         fy = finit
-         gy = ginit
-         stmin = zero
-         stmax = stp + xtrapu*stp
-         task = 'FG'
-
-         go to 10
-
-      else
-
-c        Restore local variables.
-
-         if (isave(1) .eq. 1) then
-            brackt = .true.
-         else
-            brackt = .false.
-         end if
-         stage = isave(2)
-         ginit = dsave(1)
-         gtest = dsave(2)
-         gx = dsave(3)
-         gy = dsave(4)
-         finit = dsave(5)
-         fx = dsave(6)
-         fy = dsave(7)
-         stx = dsave(8)
-         sty = dsave(9)
-         stmin = dsave(10)
-         stmax = dsave(11)
-         width = dsave(12)
-         width1 = dsave(13)
-
-      end if
-
-c     If psi(stp) <= 0 and f'(stp) >= 0 for some step, then the
-c     algorithm enters the second stage.
-
-      ftest = finit + stp*gtest
-      if (stage .eq. 1 .and. f .le. ftest .and. g .ge. zero) stage = 2
-
-c     Test for warnings.
-
-      if (brackt .and. (stp .le. stmin .or. stp .ge. stmax))
-     +    task = 'WARNING: ROUNDING ERRORS PREVENT PROGRESS'
-      if (brackt .and. stmax-stmin .le. xtol*stmax)
-     +    task = 'WARNING: XTOL TEST SATISFIED'
-      if (stp .eq. stpmax .and. f .le. ftest .and. g .le. gtest)
-     +    task = 'WARNING: STP = STPMAX'
-      if (stp .eq. stpmin .and. (f .gt. ftest .or. g .ge. gtest))
-     +    task = 'WARNING: STP = STPMIN'
-
-c     Test for convergence.
-
-      if (f .le. ftest .and. abs(g) .le. gtol*(-ginit))
-     +    task = 'CONVERGENCE'
-
-c     Test for termination.
-
-      if (task(1:4) .eq. 'WARN' .or. task(1:4) .eq. 'CONV') go to 10
-
-c     A modified function is used to predict the step during the
-c     first stage if a lower function value has been obtained but
-c     the decrease is not sufficient.
-
-      if (stage .eq. 1 .and. f .le. fx .and. f .gt. ftest) then
-
-c        Define the modified function and derivative values.
-
-         fm = f - stp*gtest
-         fxm = fx - stx*gtest
-         fym = fy - sty*gtest
-         gm = g - gtest
-         gxm = gx - gtest
-         gym = gy - gtest
-
-c        Call dcstep to update stx, sty, and to compute the new step.
-
-         call dcstep(stx,fxm,gxm,sty,fym,gym,stp,fm,gm,brackt,stmin,
-     +               stmax)
-
-c        Reset the function and derivative values for f.
-
-         fx = fxm + stx*gtest
-         fy = fym + sty*gtest
-         gx = gxm + gtest
-         gy = gym + gtest
-
-      else
-
-c       Call dcstep to update stx, sty, and to compute the new step.
-
-         call dcstep(stx,fx,gx,sty,fy,gy,stp,f,g,brackt,stmin,stmax)
-
-      end if
-
-c     Decide if a bisection step is needed.
-
-      if (brackt) then
-         if (abs(sty-stx) .ge. p66*width1) stp = stx + p5*(sty-stx)
-         width1 = width
-         width = abs(sty-stx)
-      end if
-
-c     Set the minimum and maximum steps allowed for stp.
-
-      if (brackt) then
-         stmin = min(stx,sty)
-         stmax = max(stx,sty)
-      else
-         stmin = stp + xtrapl*(stp-stx)
-         stmax = stp + xtrapu*(stp-stx)
-      end if
-
-c     Force the step to be within the bounds stpmax and stpmin.
-
-      stp = max(stp,stpmin)
-      stp = min(stp,stpmax)
-
-c     If further progress is not possible, let stp be the best
-c     point obtained during the search.
-
-      if (brackt .and. (stp .le. stmin .or. stp .ge. stmax) .or.
-     +    (brackt .and. stmax-stmin .le. xtol*stmax)) stp = stx
-
-c     Obtain another function and derivative.
-
-      task = 'FG'
-
-   10 continue
-
-c     Save local variables.
-
-      if (brackt) then
-         isave(1) = 1
-      else
-         isave(1) = 0
-      end if
-      isave(2) = stage
-      dsave(1) = ginit
-      dsave(2) = gtest
-      dsave(3) = gx
-      dsave(4) = gy
-      dsave(5) = finit
-      dsave(6) = fx
-      dsave(7) = fy
-      dsave(8) = stx
-      dsave(9) = sty
-      dsave(10) = stmin
-      dsave(11) = stmax
-      dsave(12) = width
-      dsave(13) = width1
-
-      end
diff --git a/scipy/optimize/minpack2/dcstep.f b/scipy/optimize/minpack2/dcstep.f
deleted file mode 100644
index a6c0d9e23..000000000
--- a/scipy/optimize/minpack2/dcstep.f
+++ /dev/null
@@ -1,254 +0,0 @@
-      subroutine dcstep(stx,fx,dx,sty,fy,dy,stp,fp,dp,brackt,stpmin,
-     +                  stpmax)
-      logical brackt
-      double precision stx, fx, dx, sty, fy, dy, stp, fp, dp, stpmin,
-     +                 stpmax
-c     **********
-c
-c     Subroutine dcstep
-c
-c     This subroutine computes a safeguarded step for a search
-c     procedure and updates an interval that contains a step that
-c     satisfies a sufficient decrease and a curvature condition.
-c
-c     The parameter stx contains the step with the least function
-c     value. If brackt is set to .true. then a minimizer has
-c     been bracketed in an interval with endpoints stx and sty.
-c     The parameter stp contains the current step.
-c     The subroutine assumes that if brackt is set to .true. then
-c
-c           min(stx,sty) < stp < max(stx,sty),
-c
-c     and that the derivative at stx is negative in the direction
-c     of the step.
-c
-c     The subroutine statement is
-c
-c       subroutine dcstep(stx,fx,dx,sty,fy,dy,stp,fp,dp,brackt,
-c                         stpmin,stpmax)
-c
-c     where
-c
-c       stx is a double precision variable.
-c         On entry stx is the best step obtained so far and is an
-c            endpoint of the interval that contains the minimizer.
-c         On exit stx is the updated best step.
-c
-c       fx is a double precision variable.
-c         On entry fx is the function at stx.
-c         On exit fx is the function at stx.
-c
-c       dx is a double precision variable.
-c         On entry dx is the derivative of the function at
-c            stx. The derivative must be negative in the direction of
-c            the step, that is, dx and stp - stx must have opposite
-c            signs.
-c         On exit dx is the derivative of the function at stx.
-c
-c       sty is a double precision variable.
-c         On entry sty is the second endpoint of the interval that
-c            contains the minimizer.
-c         On exit sty is the updated endpoint of the interval that
-c            contains the minimizer.
-c
-c       fy is a double precision variable.
-c         On entry fy is the function at sty.
-c         On exit fy is the function at sty.
-c
-c       dy is a double precision variable.
-c         On entry dy is the derivative of the function at sty.
-c         On exit dy is the derivative of the function at the exit sty.
-c
-c       stp is a double precision variable.
-c         On entry stp is the current step. If brackt is set to .true.
-c            then on input stp must be between stx and sty.
-c         On exit stp is a new trial step.
-c
-c       fp is a double precision variable.
-c         On entry fp is the function at stp
-c         On exit fp is unchanged.
-c
-c       dp is a double precision variable.
-c         On entry dp is the the derivative of the function at stp.
-c         On exit dp is unchanged.
-c
-c       brackt is an logical variable.
-c         On entry brackt specifies if a minimizer has been bracketed.
-c            Initially brackt must be set to .false.
-c         On exit brackt specifies if a minimizer has been bracketed.
-c            When a minimizer is bracketed brackt is set to .true.
-c
-c       stpmin is a double precision variable.
-c         On entry stpmin is a lower bound for the step.
-c         On exit stpmin is unchanged.
-c
-c       stpmax is a double precision variable.
-c         On entry stpmax is an upper bound for the step.
-c         On exit stpmax is unchanged.
-c
-c     MINPACK-1 Project. June 1983
-c     Argonne National Laboratory.
-c     Jorge J. More' and David J. Thuente.
-c
-c     MINPACK-2 Project. November 1993.
-c     Argonne National Laboratory and University of Minnesota.
-c     Brett M. Averick and Jorge J. More'.
-c
-c     **********
-      double precision zero, p66, two, three
-      parameter (zero=0.0d0,p66=0.66d0,two=2.0d0,three=3.0d0)
-
-      double precision gamma, p, q, r, s, sgnd, stpc, stpf, stpq, theta
-
-      sgnd = dp*(dx/abs(dx))
-
-c     First case: A higher function value. The minimum is bracketed.
-c     If the cubic step is closer to stx than the quadratic step, the
-c     cubic step is taken, otherwise the average of the cubic and
-c     quadratic steps is taken.
-
-      if (fp .gt. fx) then
-         theta = three*(fx-fp)/(stp-stx) + dx + dp
-         s = max(abs(theta),abs(dx),abs(dp))
-         gamma = s*sqrt((theta/s)**2-(dx/s)*(dp/s))
-         if (stp .lt. stx) gamma = -gamma
-         p = (gamma-dx) + theta
-         q = ((gamma-dx)+gamma) + dp
-         r = p/q
-         stpc = stx + r*(stp-stx)
-         stpq = stx + ((dx/((fx-fp)/(stp-stx)+dx))/two)*(stp-stx)
-         if (abs(stpc-stx) .lt. abs(stpq-stx)) then
-            stpf = stpc
-         else
-            stpf = stpc + (stpq-stpc)/two
-         end if
-         brackt = .true.
-
-c     Second case: A lower function value and derivatives of opposite
-c     sign. The minimum is bracketed. If the cubic step is farther from
-c     stp than the secant step, the cubic step is taken, otherwise the
-c     secant step is taken.
-
-      else if (sgnd .lt. zero) then
-         theta = three*(fx-fp)/(stp-stx) + dx + dp
-         s = max(abs(theta),abs(dx),abs(dp))
-         gamma = s*sqrt((theta/s)**2-(dx/s)*(dp/s))
-         if (stp .gt. stx) gamma = -gamma
-         p = (gamma-dp) + theta
-         q = ((gamma-dp)+gamma) + dx
-         r = p/q
-         stpc = stp + r*(stx-stp)
-         stpq = stp + (dp/(dp-dx))*(stx-stp)
-         if (abs(stpc-stp) .gt. abs(stpq-stp)) then
-            stpf = stpc
-         else
-            stpf = stpq
-         end if
-         brackt = .true.
-
-c     Third case: A lower function value, derivatives of the same sign,
-c     and the magnitude of the derivative decreases.
-
-      else if (abs(dp) .lt. abs(dx)) then
-
-c        The cubic step is computed only if the cubic tends to infinity
-c        in the direction of the step or if the minimum of the cubic
-c        is beyond stp. Otherwise the cubic step is defined to be the
-c        secant step.
-
-         theta = three*(fx-fp)/(stp-stx) + dx + dp
-         s = max(abs(theta),abs(dx),abs(dp))
-
-c        The case gamma = 0 only arises if the cubic does not tend
-c        to infinity in the direction of the step.
-
-         gamma = s*sqrt(max(zero,(theta/s)**2-(dx/s)*(dp/s)))
-         if (stp .gt. stx) gamma = -gamma
-         p = (gamma-dp) + theta
-         q = (gamma+(dx-dp)) + gamma
-         r = p/q
-         if (r .lt. zero .and. gamma .ne. zero) then
-            stpc = stp + r*(stx-stp)
-         else if (stp .gt. stx) then
-            stpc = stpmax
-         else
-            stpc = stpmin
-         end if
-         stpq = stp + (dp/(dp-dx))*(stx-stp)
-
-         if (brackt) then
-
-c           A minimizer has been bracketed. If the cubic step is
-c           closer to stp than the secant step, the cubic step is
-c           taken, otherwise the secant step is taken.
-
-            if (abs(stpc-stp) .lt. abs(stpq-stp)) then
-               stpf = stpc
-            else
-               stpf = stpq
-            end if
-            if (stp .gt. stx) then
-               stpf = min(stp+p66*(sty-stp),stpf)
-            else
-               stpf = max(stp+p66*(sty-stp),stpf)
-            end if
-         else
-
-c           A minimizer has not been bracketed. If the cubic step is
-c           farther from stp than the secant step, the cubic step is
-c           taken, otherwise the secant step is taken.
-
-            if (abs(stpc-stp) .gt. abs(stpq-stp)) then
-               stpf = stpc
-            else
-               stpf = stpq
-            end if
-            stpf = min(stpmax,stpf)
-            stpf = max(stpmin,stpf)
-         end if
-
-c     Fourth case: A lower function value, derivatives of the same sign,
-c     and the magnitude of the derivative does not decrease. If the
-c     minimum is not bracketed, the step is either stpmin or stpmax,
-c     otherwise the cubic step is taken.
-
-      else
-         if (brackt) then
-            theta = three*(fp-fy)/(sty-stp) + dy + dp
-            s = max(abs(theta),abs(dy),abs(dp))
-            gamma = s*sqrt((theta/s)**2-(dy/s)*(dp/s))
-            if (stp .gt. sty) gamma = -gamma
-            p = (gamma-dp) + theta
-            q = ((gamma-dp)+gamma) + dy
-            r = p/q
-            stpc = stp + r*(sty-stp)
-            stpf = stpc
-         else if (stp .gt. stx) then
-            stpf = stpmax
-         else
-            stpf = stpmin
-         end if
-      end if
-
-c     Update the interval which contains a minimizer.
-
-      if (fp .gt. fx) then
-         sty = stp
-         fy = fp
-         dy = dp
-      else
-         if (sgnd .lt. zero) then
-            sty = stx
-            fy = fx
-            dy = dx
-         end if
-         stx = stp
-         fx = fp
-         dx = dp
-      end if
-
-c     Compute the new step.
-
-      stp = stpf
-
-      end
diff --git a/scipy/optimize/minpack2/minpack2.pyf b/scipy/optimize/minpack2/minpack2.pyf
deleted file mode 100644
index b900edcca..000000000
--- a/scipy/optimize/minpack2/minpack2.pyf
+++ /dev/null
@@ -1,35 +0,0 @@
-!    -*- f90 -*-
-python module _minpack2 ! in 
-    interface  ! in :minpack2
-        subroutine dcsrch(stp,f,g,ftol,gtol,xtol,task,stpmin,stpmax,isave,dsave) ! in :minpack2:dcsrch.f
-            double precision, intent(in,out) :: stp
-            double precision, intent(in,out) :: f
-            double precision, intent(in,out) :: g
-            double precision, intent(in) :: ftol
-            double precision, intent(in) :: gtol
-            double precision, intent(in) :: xtol
-            character*60, intent(in, out) :: task
-            double precision, intent(in) :: stpmin
-            double precision, intent(in) :: stpmax
-            integer dimension(2), intent(inout) :: isave
-            double precision dimension(13), intent(inout) :: dsave
-        end subroutine dcsrch
-        subroutine dcstep(stx,fx,dx,sty,fy,dy,stp,fp,dp,brackt,stpmin,stpmax) ! in :minpack2:dcstep.f
-            double precision, intent(in,out) :: stx
-            double precision, intent(in,out) :: fx
-            double precision, intent(in,out) :: dx
-            double precision, intent(in,out) :: sty
-            double precision, intent(in,out) :: fy
-            double precision, intent(in,out) :: dy
-            double precision, intent(in,out) :: stp
-            double precision :: fp
-            double precision :: dp
-            logical, intent(in,out) :: brackt
-            double precision :: stpmin
-            double precision :: stpmax
-        end subroutine dcstep
-    end interface 
-end python module minpack2
-
-! This file was auto-generated with f2py (version:2.39.235_1703).
-! See http://cens.ioc.ee/projects/f2py2e/
diff --git a/scipy/optimize/setup.py b/scipy/optimize/setup.py
index 850f73948..670f8f750 100644
--- a/scipy/optimize/setup.py
+++ b/scipy/optimize/setup.py
@@ -84,10 +84,6 @@ def configuration(parent_package='', top_path=None):
                              'cobyla.pyf', 'cobyla2.f', 'trstlp.f']],
                          **numpy_nodepr_api)
 
-    sources = ['minpack2.pyf', 'dcsrch.f', 'dcstep.f']
-    config.add_extension('_minpack2',
-                         sources=[join('minpack2', x) for x in sources],
-                         **numpy_nodepr_api)
 
     sources = ['slsqp.pyf', 'slsqp_optmz.f']
     ext = config.add_extension('_slsqp', sources=[
```
</p>
</details>

andyfaff commented:
@j-bowhay I've dropped the commit that removed all of the Fortran dcsrch. The patch is above for when we decide it can be removed.

j-bowhay reviewed (approved):
Thanks @andyfaff LGTM subject to CI. I will wait around a week before merging in case anyone else wants to look.

ilayn commented:
I think after two versions we can safely remove the Fortran code for this.

j-bowhay commented:
> I think after two versions we can safely remove the Fortran code for this.

Agreed, the public interface to this got deprecated anyway so we are completely safe to remove, I'll open a pr
